File: Programmers/HashMap/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

int solution(vector<string> friends, vector<string> gifts) {
    unordered_map<string, int> hm;
    int n = friends.size();
    vector<vector<int>> graph (n, vector<int>(n));
    vector<int> give(n), receive(n);
    vector<int> score(n);
    for (int i = 0; i < friends.size(); i++) {
        hm[friends[i]] = i;
    }

    for (auto gift : gifts) {
        int idx = gift.find(' ');
        string a = gift.substr(0, idx);
        string b = gift.substr(idx + 1);

        int from_idx = hm[a];
        int to_idx = hm[b];
        graph[from_idx][to_idx]++;
        give[from_idx]++;
        receive[to_idx]++;
    }

    for (int i = 0; i < n; i++) {
        score[i] = give[i] - receive[i];
    }

    vector<int> result(n);

    for (int i = 0 ; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i == j) continue;
            if (graph[i][j] > graph[j][i]) {
                result[i]++;
            } else if (graph[i][j] == graph[j][i]) {
                if (score[i] > score[j]) result[i]++;
            }
        }
    }

    int answer = 0;
    for (int i = 0; i < n; i++) {
        answer = max(answer, result[i]);
    }
    return answer;
}

File: Programmers/구현/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    n = len(friends)
    pair = [[0] * n for _ in range(n)]
    give = [0] * n
    recv = [0] * n
    friend = {name: i for i, name in enumerate(friends)}
    print(friend)

    for gift in gifts:
        a, b = gift.split()
        i, j = friend[a], friend[b]
        pair[i][j] += 1
        give[i] += 1
        recv[j] += 1

    next_recv = [0] * n
    score = [give[i] - recv[i] for i in range(n)]
    for i in range(n):
        for j in range(n):
            if i == j: continue
            if pair[i][j] > pair[j][i]:
                next_recv[i] += 1
            elif pair[i][j] == pair[j][i]:
                if score[i] > score[j]:
                    next_recv[i] += 1

    return max(next_recv) if next_recv else 0